packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_def.hpp

```diff
@@ -411,74 +403,114 @@ struct UnpackCrsMatrixAndCombineFunctor {
 
 };
 
-/// \brief Functor to determine the maximum number of entries in any row of the
-///   matrix using Kokkos::parallel_reduce
+struct MaxNumEntTag {};
+struct TotNumEntTag {};
+
+/// \brief Functor to determine the number of entries in a matrix using
+/// Kokkos::parallel_reduce
```

Comment:
The documentation looks wrong -- this is a functor to determine either the total  number of bytes required for unpacking incoming matrix entries, or the maximum number of bytes required for unpacking a row of incoming matrix entries, right?

Reply:
If so, could you do both at once?  Each requires one pass over the incoming packed data, so you could get the max and total in one shot.

Reply:
I believe the fun for does compute the max and total number of entries for array allocation purposes.  They can probably be combined

Reply:
The tag could be more explicit, it's the max bum entries on any one row.  It's used to allocate scratch space on device

Reply:
@mhoemmen said:

> if so, could you do both at once? Each requires one pass over the incoming packed data, so you could get the max and total in one shot.

I suppose they could be combined, but it would not save work done as the two counting functions are not called in the same procedure.  Unpacking in to the `CrsMatrix` requires the maximum number of entries in any one row and unpacking in to `CrsArrays`  requires total number of entries, thus two separate counting procedures (but only one functor)

Reply:
> but it would not save work done as the two counting functions are not called in the same procedure.

It seems like you could compute both of them once, before anybody needs them, and then pass along those two integer values (the max and the total).

Reply:
Maybe I wasn't too clear.  The total is needed for `transferAndFillComplete` and the maximum is needed for `packAndPrepare`.  I'm having a hard time seeing how computing the max and total at once would be beneficial as these two procedures are independent.